!htt #19388 Removing use of InetAddress.getByName

cmbaxter opened the pull request:
- Remove String apply methods for RemoteAddress (both scala and java)
- Added unit test to show that hostnames in the address are treated as invalid by the parser
- Added a unit test to show that invalid xff is added as a RawHeader instead of model class

cmbaxter committed:
!htt #19388 Removing use of InetAddress.getByName

* Remove String apply methods for RemoteAddress (both scala and java)
* Added unit test to show that hostnames in the address are treated as invalid by the parser
* Added a unit test to show that invalid xff is added as a RawHeader instead of model class

cmbaxter committed:
!htt #19388 Removing use of InetAddress.getByName

* Remove String apply methods for RemoteAddress (both scala and java)
* Added unit test to show that hostnames in the address are treated as invalid by the parser
* Added a unit test to show that invalid xff is added as a RawHeader instead of model class

ktoso commented:
OK TO TEST

ktoso commented:
OK TO TEST

drewhk commented:
Refs #19388

drewhk commented:
LGTM, but might need a mention in the migration guide.

drewhk commented:
This does not compile.

cmbaxter commented:
@drewhk, ya, I noticed that when the build failed about a week ago.  I then went back to the original ticket and commented as to why.  Can you check out my comments and let me know if you have any feedback before I proceed?

https://github.com/akka/akka/issues/19388

drewhk commented:
I think the approach you outlined in the ticket is the good one. There is only one thing to consider here, is that we will be cutting a 2.4.2 soon (but first an RC2 so not *too* soon) and after that we cannot make binary incompatible changes. So this needs to go in this week. We can take it over if you prefer.

drewhk commented:
@cmbaxter do you want us to take over this ticket?

cmbaxter commented:
@drewhk, I apologize again for not getting on this sooner.  Gotta pay the bills with the day job first :)

I want to take a stab at this again right now.  I need to refresh my feature branch from master and then see how much more work it's going to be to fix the logic from x-real-ip.  If I can't get this fully done today then I will pass the torch to you guys.  I'll write back shortly letting you know when I am with this.  Hopefully I will have a new PR done for today.

cmbaxter commented:
@drewhk, when I pulled down the latest master and integrated it into my feature branch, I see a test in akka-http-core is currently failing.  It's the `	akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec` spec.  I don't think this has anything to do with my changes, but this test either didn't exist before or wasn't failing before.  Do I need to be concerned with this before moving forward to fixing the issue with x-real-ip?

ktoso commented:
hey @cmbaxter, sorry for the failure on master; 
@johanandren has a fix for it already. issue: https://github.com/akka/akka/issues/19741 pr: https://github.com/akka/akka/pull/19742

drewhk commented:
That is a know failure that somehow sneaked in. we have a fix pending. Once we merged that we can rerun the build here. Until then ignore that failure.

cmbaxter commented:
@drewhk, @ktoso, okay, I will ignore the warning.

For the code impl, my suggestion was to use the code in the `IpAddressParsing` trait to vet the raw x-real-ip header and get the bytes for if it valid.  I see two possible approaches here.  Let me know which of these you prefer:

* Add an `object IpAddressParsing` that extends that trait and has a method to vet a string ip and returns an `Option[Array[Byte]]` depending on if the address is valid according to the rules.  That trait is currently package private and this object would have to not be so that I can then leverage if in the `extractClientIp` directive to properly vet that raw header value from x-real-ip.  Probably not ideal to have to expose this functionality that is currently intentionally package private, but this is the easier thing to do.
* Setup`X-Real-Ip` as a properly modeled header (no longer a raw header) and then hook the rules into the parsing (similar to the `Remote-Address` header).  This takes more work but doesn't require exposing that ip vetting logic outside of that `parsing` package.

If there are other better options please let me know.  I want to get this done right first without having to go back and forth on code reviews based on the PR.

drewhk commented:
I personally prefer the second option (proper header).

cmbaxter committed:
!htt #19388 Chanring x-real-ip from raw to real modeled header

* Allows for parsing rules to be applied to vet the ip
* Removed last cases of code where the String variant of apply on RemoteAddress is used

cmbaxter committed:
=htt #19388 Fixing broken test in akka-docs from RemoteAddress changes

Also dealing with merge conflicts from my fork's head

Conflicts:
	akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala
	akka-http-core/src/test/scala/akka/http/impl/model/parser/HttpHeaderSpec.scala

cmbaxter commented:
@drewhk, okay, changes have been committed and the PR amended to contain them.  I'll be watching for the CI run whenever that kicks off.

drewhk commented:
Argh, something is wrong, there are unrelated comments in this PR!

cmbaxter commented:
@drewhk, there was an actual compilation failure in akka-docs, on a test case in there that used the old String apply from RemoteAddress.  I have that fixed and am ready to commit that fix as well.  Is there something else wrong?  I noticed one of my commits, which was we merging from the head of my fork's feature branch does not comply.  Do I need to fix that too?

cmbaxter commented:
Okay, so the only test failure there was the previously mentioned `akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec`.  Am I in the clear here?

cmbaxter commented:
@drewhk, are we good with this PR considering the test fails were not related to my changes?

drewhk commented:
Ok, so now we have proper parser rules, so the DNS problem is no longer there. LGTM from my side, but we will need a second reviewer. We can take over from here if modifications are needed.

ktoso commented:
LGTM, thanks a lot! The failure is known and short term wise has been addressed, long term we're working on it (it's related to OutputTruncated)
